storagenode/retain/retain.go

```diff
@@ -88,105 +89,207 @@ type Service struct {
 // true is returned if the request is queued and false is returned if it is discarded
 func (s *Service) Queue(req Request) bool {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.cond.L.Lock()
+	defer s.cond.L.Unlock()
 
-	// subtract some time to leave room for clock difference between the satellite and storage node
-	req.CreatedBefore = req.CreatedBefore.Add(-s.config.RetainTimeBuffer)
+	select {
+	case <-s.closed:
+		return false
+	default:
+	}
 
-	// only queue retain request if we do not already have one for this satellite
-	if _, ok := s.queued[req.SatelliteID]; !ok {
-		s.queued[req.SatelliteID] = req
-		go func() { s.reqChan <- req }()
+	// subtract some time to leave room for clock difference between the satellite and storage node
+	req.CreatedBefore = req.CreatedBefore.Add(-s.config.MaxTimeSkew)
 
-		return true
-	}
+	s.queued[req.SatelliteID] = req
+	s.cond.Broadcast()
 
-	return false
+	return true
 }
 
 // Run listens for queued retain requests and processes them as they come in.
 func (s *Service) Run(ctx context.Context) error {
-	for {
-		// exit if context has been canceled. Otherwise, block until an item can be added to the semaphore
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case s.semaphore <- struct{}{}:
-		}
+	// Hold the lock while we spawn the workers because a concurrent Close call
+	// can race and wait for them.
+	s.cond.L.Lock()
+	defer s.cond.L.Unlock()
+
+	// Ensure Run is only ever called once. If not, there's many subtle
+	// bugs with concurrency.
+	if s.started {
+		return Error.New("service already started")
+	}
+	s.started = true
+
+	// Ensure Run doesn't start after it's closed. Then we may leak some
+	// workers.
+	select {
+	case <-s.closed:
+		return Error.New("service Closed")
+	default:
+	}
+
+	// Create a sub-context that we can cancel.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Start a goroutine that does most of the work of Close in the case the
+	// context is canceled and broadcasts to anyone waiting on the condition
+	// variable to check the state.
+	s.group.Go(func() error {
+		defer s.cond.Broadcast()
+		defer cancel()
 
-		// get the next request
-		var req Request
 		select {
-		case req = <-s.reqChan:
+		case <-s.closed:
+			return nil
+
 		case <-ctx.Done():
+			s.cond.L.Lock()
+			s.closedOnce.Do(func() { close(s.closed) })
+			s.cond.L.Unlock()
+
 			return ctx.Err()
 		}
+	})
 
-		go func(ctx context.Context, req Request) {
-			err := s.retainPieces(ctx, req)
-			if err != nil {
-				s.log.Error("retain error", zap.Error(err))
-			}
-			s.mu.Lock()
-			delete(s.queued, req.SatelliteID)
-			queueLength := len(s.queued)
-			s.mu.Unlock()
+	for i := 0; i < s.config.Concurrency; i++ {
+		s.group.Go(func() error {
+			// Grab lock to check things.
+			s.cond.L.Lock()
+			defer s.cond.L.Unlock()
+
+			for {
+				// If we have closed, exit.
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case <-s.closed:
```

Comment:
Only checking s.closed would suffice, but there's no harm in checking both. The context cancellation error can be returned outside.